torch/csrc/distributed/autograd/context/dist_autograd_context.cpp

```diff
@@ -48,6 +48,68 @@ DistAutogradContext::recvFunctions() const {
   return recvAutogradFunctions_;
 }
 
+void DistAutogradContext::accumulateGrad(
+    torch::autograd::Variable variable,
+    torch::Tensor grad) {
+  TORCH_INTERNAL_ASSERT(grad.defined());
+  TORCH_INTERNAL_ASSERT(variable.requires_grad());
+
+  std::lock_guard<std::mutex> guard(lock_);
+  auto it = accumulatedGrads_.find(variable);
+  if (it != accumulatedGrads_.end()) {
+    // Accumulate multiple grads on the same variable.
+    it->setValue(it->value() + grad);
+  } else {
+    // First grad for this variable.
+    accumulatedGrads_.insert(variable, grad);
+  }
+}
+
+torch::autograd::GraphTask& DistAutogradContext::retrieveGraphTask() {
+  std::lock_guard<std::mutex> guard(lock_);
+  TORCH_INTERNAL_ASSERT(graphTask_);
+  return *graphTask_;
+}
+
+void DistAutogradContext::setGraphTask(
+    std::unique_ptr<torch::autograd::GraphTask> graphTask) {
+  std::lock_guard<std::mutex> guard(lock_);
+  graphTask_ = std::move(graphTask);
+}
+
+void DistAutogradContext::addOutStandingRpc(
+    std::shared_ptr<rpc::FutureMessage> futureMessage) {
+  std::lock_guard<std::mutex> guard(lock_);
+  outStandingRpcs_.push_back(futureMessage);
+}
+
+void DistAutogradContext::waitForOutStandingRpcs() {
+  // Copy futures under lock, but wait for them outside the lock.
```

Comment:
Why not move them out? (Since after you wait, they will all never need waiting in the future!)

Reply:
(I suppose if you make concurrent calls here, you need to wait for the outstanding rpcs (which you wouldn't if you cleared `outStandingRpcs_`)) But then it feels like you should clear it after you're done waiting.